src/controller/proxy/controller.go

```diff
@@ -0,0 +1,224 @@
+//  Copyright Project Harbor Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/controller/artifact"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/model"
+	"github.com/goharbor/harbor/src/replication/registry"
+	"github.com/opencontainers/go-digest"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	maxWait             = 10
+	manifestListWaitSec = 1800
+	sleepIntervalSec    = 20
+)
+
+var (
+	// Ctl is a global proxy controller instance
+	Ctl = NewController()
+)
+
+// Controller defines the operations related with pull through proxy
+type Controller interface {
+	// UseLocalManifest check if the manifest should use local
+	UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool
+	// UseLocalBlob check if the blob should use local
+	UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool
+	// ProxyBlob proxy the blob request to the target server
+	ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error
+}
+
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	mu          sync.Mutex
+	inflight    map[string]interface{}
+	artifactCtl artifact.Controller
+	local       LocalInterface
+}
+
+func (c *controller) isProxyReady(p *models.Project) bool {
+	if p.RegistryID < 1 {
+		return false
+	}
+	reg, err := c.registryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Errorf("failed to get registry, error:%v", err)
+		return false
+	}
+	return reg.Status == model.Healthy
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool {
+	if !c.isProxyReady(p) {
```

Comment:
The checking is useless? We can just check the existence in local?

Reply:
Move to the middleware